test/test_gitreceive.go

```diff
@@ -53,3 +60,63 @@ func (s *GitreceiveSuite) TestSlugbuilderLimit(t *c.C) {
 
 	t.Assert(r.flynn("-a", "gitreceive", "env", "unset", "SLUGBUILDER_DEFAULT_MEMORY_LIMIT"), Succeeds)
 }
+
+func (s *GitreceiveSuite) deployTar(t *c.C, app *ct.App, rev string, env, meta map[string]string, data io.Reader) {
+	client := s.controllerClient(t)
+
+	gitreceiveRelease, err := client.GetAppRelease("gitreceive")
+	t.Assert(err, c.IsNil)
+
+	args := []string{
+		"/bin/flynn-receiver",
+		app.Name,
+		rev,
+	}
+
+	for name, m := range map[string]map[string]string{"--env": env, "--meta": meta} {
+		for k, v := range m {
+			args = append(args, name)
+			args = append(args, fmt.Sprintf("%s=%s", k, v))
+		}
+	}
+
+	rwc, err := client.RunJobAttached("gitreceive", &ct.NewJob{
+		ReleaseID:  gitreceiveRelease.ID,
+		ReleaseEnv: true,
+		Args:       args,
+	})
+	t.Assert(err, c.IsNil)
+	attachClient := cluster.NewAttachClient(rwc)
+	var outBuf bytes.Buffer
+	_, err = io.Copy(attachClient, data)
+	t.Assert(err, c.IsNil)
+	t.Assert(attachClient.CloseWrite(), c.IsNil)
+	exit, err := attachClient.Receive(&outBuf, &outBuf)
+	t.Log(outBuf.String())
+	if err != nil {
+		debugf(t, "error running flynn-receiver: %s", err)
+	}
+	t.Assert(exit, c.Equals, 0)
+	t.Assert(err, c.IsNil)
+}
+
+func (s *GitreceiveSuite) TestDeployWithEnv(t *c.C) {
+	r := s.newGitRepo(t, "ruby-compile-env")
+
+	client := s.controllerClient(t)
+	app := &ct.App{}
+	t.Assert(client.CreateApp(app), c.IsNil)
+	debugf(t, "created app %s (%s)", app.Name, app.ID)
+
+	data, err := os.Open(filepath.Join(r.dir, "files.tar"))
+	t.Assert(err, c.IsNil)
+	defer data.Close()
+
+	env := map[string]string{
+		"FOO": "BAR",
+	}
+	meta := map[string]string{
+		"git": "true",
+	}
+	s.deployTar(t, app, "test-rev", env, meta, data)
```

Comment:
You could simplify this test by using `tar` and the CLI from a shell:

```
tar c -C test/apps/env-dir . | flynn -a gitreceive run /bin/flynn-receiver APP REV -e FOO=BAR
```

If you do go with that, you'll probably want to use `bash -o pipefail -e -c "command"` to make sure it exits non-zero on failure.